File: python/ch15/assignment_relax.py
# python/ch15/assignment_relax.py
"""割当問題: 分数緩和 + 丸め、n=4 で全探索と比較."""
from __future__ import annotations

import itertools
import sys
from typing import List, Tuple


def read_cost(stream) -> Tuple[int, List[List[int]]]:
    lines = stream.read().splitlines()
    n = int(lines[0])
    mat = [list(map(int, lines[i + 1].split())) for i in range(n)]
    return n, mat


def fractional_relaxation(cost: List[List[int]]) -> Tuple[List[List[float]], float]:
    n = len(cost)
    entries = [(cost[i][j], i, j) for i in range(n) for j in range(n)]
    entries.sort()
    row_rem = [1.0] * n
    col_rem = [1.0] * n
    x = [[0.0] * n for _ in range(n)]
    total = 0.0
    for c, i, j in entries:
        val = min(row_rem[i], col_rem[j])
        if val <= 0:
            continue
        x[i][j] = val
        row_rem[i] -= val
        col_rem[j] -= val
        total += c * val
    return x, total


def round_assignment(cost: List[List[int]], frac: List[List[float]]) -> Tuple[List[int], int]:
    n = len(cost)
    items = [(frac[i][j], cost[i][j], i, j) for i in range(n) for j in range(n)]
    items.sort(reverse=True)
    assign = [-1] * n
    used_col = [False] * n
    total = 0
    for _, c, i, j in items:
        if assign[i] != -1 or used_col[j]:
            continue
        assign[i] = j
        used_col[j] = True
        total += c
    return assign, total


def brute_optimal(cost: List[List[int]]) -> Tuple[List[int], int]:
    n = len(cost)
    best_perm = None
    best_val = None
    for perm in itertools.permutations(range(n)):
        val = sum(cost[i][perm[i]] for i in range(n))
        if best_val is None or val < best_val:
            best_val = val
            best_perm = list(perm)
    assert best_perm is not None and best_val is not None
    return best_perm, best_val


def main() -> None:
    n, cost = read_cost(sys.stdin)
    frac, frac_cost = fractional_relaxation(cost)
    assign, rounded_cost = round_assignment(cost, frac)
    optimal, opt_cost = brute_optimal(cost)
    gap = rounded_cost - opt_cost
    print("fractional_cost", f"{frac_cost:.2f}")
    print("rounded_assign", assign, "cost", rounded_cost)
    print("optimal_assign", optimal, "cost", opt_cost)
    print(f"n={n} gap={gap} ratio={rounded_cost / max(opt_cost, 1):.3f}")


if __name__ == "__main__":
    main()

File: go/ch15/assignment_relax.go
// go/ch15/assignment_relax.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func readCost(sc *bufio.Scanner) (int, [][]int, error) {
	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		return 0, nil, err
	}
	n, _ := strconv.Atoi(lines[0])
	cost := make([][]int, n)
	for i := 0; i < n; i++ {
		row := strings.Fields(lines[i+1])
		cost[i] = make([]int, n)
		for j, s := range row {
			cost[i][j], _ = strconv.Atoi(s)
		}
	}
	return n, cost, nil
}

type fracEntry struct {
	c, i, j int
}

func fractionalRelaxation(cost [][]int) ([][]float64, float64) {
	n := len(cost)
	var entries []fracEntry
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			entries = append(entries, fracEntry{cost[i][j], i, j})
		}
	}
	sort.Slice(entries, func(a, b int) bool { return entries[a].c < entries[b].c })
	rowRem := make([]float64, n)
	colRem := make([]float64, n)
	for i := range rowRem {
		rowRem[i], colRem[i] = 1, 1
	}
	x := make([][]float64, n)
	for i := range x {
		x[i] = make([]float64, n)
	}
	total := 0.0
	for _, e := range entries {
		val := rowRem[e.i]
		if colRem[e.j] < val {
			val = colRem[e.j]
		}
		if val <= 0 {
			continue
		}
		x[e.i][e.j] = val
		rowRem[e.i] -= val
		colRem[e.j] -= val
		total += float64(e.c) * val
	}
	return x, total
}

type roundItem struct {
	frac float64
	c    int
	i, j int
}

func roundAssignment(cost [][]int, frac [][]float64) ([]int, int) {
	n := len(cost)
	var items []roundItem
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			items = append(items, roundItem{frac[i][j], cost[i][j], i, j})
		}
	}
	sort.Slice(items, func(a, b int) bool { return items[a].frac > items[b].frac })
	assign := make([]int, n)
	for i := range assign {
		assign[i] = -1
	}
	usedCol := make([]bool, n)
	total := 0
	for _, it := range items {
		if assign[it.i] != -1 || usedCol[it.j] {
			continue
		}
		assign[it.i] = it.j
		usedCol[it.j] = true
		total += it.c
	}
	return assign, total
}

func permCost(cost [][]int, perm []int) int {
	s := 0
	for i, j := range perm {
		s += cost[i][j]
	}
	return s
}

func bruteOptimal(cost [][]int) ([]int, int) {
	n := len(cost)
	perm := make([]int, n)
	for i := range perm {
		perm[i] = i
	}
	best := append([]int(nil), perm...)
	bestVal := permCost(cost, perm)
	var dfs func(int)
	dfs = func(pos int) {
		if pos == n {
			v := permCost(cost, perm)
			if v < bestVal {
				bestVal = v
				copy(best, perm)
			}
			return
		}
		for j := pos; j < n; j++ {
			perm[pos], perm[j] = perm[j], perm[pos]
			dfs(pos + 1)
			perm[pos], perm[j] = perm[j], perm[pos]
		}
	}
	dfs(0)
	return best, bestVal
}

func main() {
	_, cost, err := readCost(bufio.NewScanner(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	n := len(cost)
	frac, fracCost := fractionalRelaxation(cost)
	assign, rounded := roundAssignment(cost, frac)
	opt, optCost := bruteOptimal(cost)
	gap := rounded - optCost
	fmt.Printf("fractional_cost %.2f\n", fracCost)
	fmt.Printf("rounded_assign %v cost %d\n", assign, rounded)
	fmt.Printf("optimal_assign %v cost %d\n", opt, optCost)
	fmt.Printf("n=%d gap=%d ratio=%.3f\n", n, gap, float64(rounded)/float64(max(optCost, 1)))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
